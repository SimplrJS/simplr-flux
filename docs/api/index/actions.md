# Actions

## Classes

### [SynchronizeMapStoreAction][ClassDeclaration-5]


----------

### [MapStoreCleanUpAction][ClassDeclaration-8]


[NamespaceImport-0]: actions.md#actions
[ClassDeclaration-5]: actions/synchronizemapstoreaction.md#synchronizemapstoreaction
[ClassDeclaration-8]: actions/mapstorecleanupaction.md#mapstorecleanupaction