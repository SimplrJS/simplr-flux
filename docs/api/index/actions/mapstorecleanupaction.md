# Table of contents

* [MapStoreCleanUpAction][ClassDeclaration-8]
    * Constructor
        * [constructor()][Constructor-6]
    * Properties
        * [storeId][GetAccessor-5]

# MapStoreCleanUpAction

```typescript
class MapStoreCleanUpAction
```
## Constructor

### constructor()

```typescript
public constructor();
```

## Properties

### storeId

```typescript
public get storeId: string;
```

**Type**

string

[ClassDeclaration-8]: mapstorecleanupaction.md#mapstorecleanupaction
[Constructor-6]: mapstorecleanupaction.md#constructor
[GetAccessor-5]: mapstorecleanupaction.md#storeid