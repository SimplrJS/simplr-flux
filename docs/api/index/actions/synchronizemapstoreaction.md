# Table of contents

* [SynchronizeMapStoreAction][ClassDeclaration-5]
    * Constructor
        * [constructor()][Constructor-3]
    * Properties
        * [storeDispatchToken][GetAccessor-3]

# SynchronizeMapStoreAction

```typescript
class SynchronizeMapStoreAction
```
## Constructor

### constructor()

```typescript
public constructor();
```

## Properties

### storeDispatchToken

```typescript
public get storeDispatchToken: string;
```

**Type**

string

[ClassDeclaration-5]: synchronizemapstoreaction.md#synchronizemapstoreaction
[Constructor-3]: synchronizemapstoreaction.md#constructor
[GetAccessor-3]: synchronizemapstoreaction.md#storedispatchtoken