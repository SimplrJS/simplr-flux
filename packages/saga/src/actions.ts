// tslint:disable:no-any

/**
 * @see @url https://github.com/redux-utilities/flux-standard-action
 */
export interface FluxStandardAction<TPayload = any, TMeta = any> {
    /**
     * The `type` of an action identifies to the consumer the nature of the action that has occurred.
     * Two actions with the same `type` MUST be strictly equivalent (using `===`)
     */
    type: string;
    /**
     * The optional `payload` property MAY be any type of value.
     * It represents the payload of the action.
     * Any information about the action that is not the type or status of the action should be part of the `payload` field.
     * By convention, if `error` is `true`, the `payload` SHOULD be an error object.
     * This is akin to rejecting a promise with an error object.
     */
    payload: TPayload;
    /**
     * The optional `error` property MAY be set to true if the action represents an error.
     * An action whose `error` is true is analogous to a rejected Promise.
     * By convention, the `payload` SHOULD be an error object.
     * If `error` has any other value besides `true`, including `undefined`, the action MUST NOT be interpreted as an error.
     */
    error?: boolean;
    /**
     * The optional `meta` property MAY be any type of value.
     * It is intended for any extra information that is not part of the payload.
     */
    meta?: TMeta;
}

export interface ErrorFluxStandardAction<TCustomError extends Error, TMeta = undefined> extends FluxStandardAction<TCustomError, TMeta> {
    error: true;
}

/**
 * Alias for FluxStandardAction.
 */
export type FSA<TPayload = any, TMeta = any> = FluxStandardAction<TPayload, TMeta>;

/**
 * Alias for ErrorFluxStandardAction.
 */
export type ErrorFSA<TCustomError extends Error, TMeta = undefined> = ErrorFluxStandardAction<TCustomError, TMeta>;

export function createAction<TAction extends FSA, TMeta = undefined>(
    type: TAction["type"],
    payload: TAction["payload"],
    meta?: TMeta
): TAction {
    if (payload != null && typeof payload !== "object") {
        throw new Error("createAction(...): Payload can only be object or undefined/null.");
    }

    const isError = (payload as object) instanceof Error;

    return {
        type: type,
        payload: payload,
        error: isError,
        meta: meta
    } as TAction;
}

/**
 * Returns `true` if `action` is FSA compliant.
 */
export function isAction<TPayload, TMeta = undefined>(action: any): action is FSA<TPayload, TMeta> {
    return typeof action === "object" && typeof action.type === "string" && Object.keys(action).every(isValidKey);
}

/**
 * Returns `true` if `action` is FSA compliant error.
 */
export function isErrorAction<TCustomError extends Error, TMeta = undefined>(action: any): action is ErrorFSA<TCustomError, TMeta> {
    return isAction(action) && action.error === true;
}

function isValidKey(key: string): boolean {
    return ["type", "payload", "error", "meta"].indexOf(key) > -1;
}
