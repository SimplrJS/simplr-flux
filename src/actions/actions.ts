export class DataMapStoreUpdatedAction {
    constructor(private storeId: string) { }
    public get StoreId() {
        return this.storeId;
    }
}

export class DataMapStoreCleanUpAction {
    constructor(private storeId: string) { }
    public get StoreId() {
        return this.storeId;
    }
}
