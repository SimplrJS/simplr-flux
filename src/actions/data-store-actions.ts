export class StoreUpdatedAction {
    constructor(private _dispatchToken: string) { }
    public get dispatchToken(): string {
        return this._dispatchToken;
    }
}
