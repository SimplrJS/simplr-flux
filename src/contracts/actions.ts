export interface FluxAction {
    type: string;
}

export interface SimplrAction<TAction = any> extends FluxAction {
    type: "SIMPLR_ACTION";
    action: TAction;
}

export type DispatcherMessage<TAction = any> = FluxAction | SimplrAction<TAction>;
