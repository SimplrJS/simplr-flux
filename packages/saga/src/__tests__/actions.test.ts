import { createSagaAction, createFluxAction, FSA } from "../actions";

jest.mock("../helpers", () => ({
    ...jest.requireActual("../helpers"),
    generateRandomString: () => "RANDOM"
}));

class TestSagaAction {}

interface TestFSA extends FSA {
    type: "TEST";
}

it("creates saga action from class payload", () => {
    expect(createSagaAction(TestSagaAction)).toMatchSnapshot();
});

it("throw when creating saga action from object payload", () => {
    expect(() => createSagaAction({})).toThrow();
});

it("creates FSA action", () => {
    expect(createFluxAction<TestFSA>("TEST", undefined)).toMatchSnapshot();
});
