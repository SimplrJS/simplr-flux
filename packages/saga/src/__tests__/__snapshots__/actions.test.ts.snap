// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates FSA action 1`] = `
Object {
  "error": false,
  "meta": undefined,
  "payload": undefined,
  "type": "TEST",
}
`;

exports[`creates saga action from class payload 1`] = `
Object {
  "error": false,
  "meta": undefined,
  "payload": [Function],
  "type": "SAGA_RANDOM",
}
`;
